Badger storage factory panics during Close

**Describe the bug**

The Badger storage factory has a inverted dependency that causes the factory to be closed twice, which leads to 
```
panic: close of closed channel

goroutine 1 [running]:
github.com/jaegertracing/jaeger/plugin/storage/badger.(*Factory).Close(0xc000692000, 0x5388600, 0xc000692000)
	/Users/ysh/dev/jaegertracing/jaeger/plugin/storage/badger/factory.go:175 +0x5b
github.com/jaegertracing/jaeger/plugin/storage.(*Factory).Close(0xc00030e000, 0x0, 0x0)
	/Users/ysh/dev/jaegertracing/jaeger/plugin/storage/factory.go:246 +0x19d
main.main.func1.1()
```

**To Reproduce**

1. Run `$ SPAN_STORAGE_TYPE=badger go run ./cmd/all-in-one`
2. Once it starts, kill process with Ctrl-C

**Expected behavior**
Process should exit cleanly.

**Version (please complete the following information):**
 - OS: MacOS
 - Jaeger version: 1.21
 - Deployment: run from source

**Additional context**

The underlying cause is that when Badger Factory creates SpanWriter, it passes itself as a `closer` argument, which results in the following sequence:
1. main -> spanWriter.Close() -> factory.Close()
2. main -> factory.Close() -> panic

**Proposed Solution**

It is not clear why the factory needs to be passed into SpanWriter, it's an inversion of control. SpanWriter.Close() should do the clean-up that it needs to do, but then expect the parent to call Factory.Close() explicitly (which is what main does).

--- a/plugin/storage/badger/dependencystore/storage_test.go
+++ b/plugin/storage/badger/dependencystore/storage_test.go
@@ -59,7 +59,7 @@ func runFactoryTest(tb testing.TB, test func(tb testing.TB, sw spanstore.Writer,
 		} else {
 			tb.FailNow()
 		}
-
+		assert.NoError(tb, f.Close())
 	}()
 	test(tb, sw, dr)
 }

--- a/plugin/storage/badger/dependencystore/storage_test.go
+++ b/plugin/storage/badger/dependencystore/storage_test.go
@@ -17,7 +17,6 @@ package dependencystore_test
 import (
 	"context"
 	"fmt"
-	"io"
 	"testing"
 	"time"
 
@@ -35,6 +34,10 @@ import (
 // Opens a badger db and runs a test on it.
 func runFactoryTest(tb testing.TB, test func(tb testing.TB, sw spanstore.Writer, dr dependencystore.Reader)) {
 	f := badger.NewFactory()
+	defer func() {
+		assert.NoError(tb, f.Close())
+	}()
+
 	opts := badger.NewOptions("badger")
 	v, command := config.Viperize(opts.AddFlags)
 	command.ParseFlags([]string{
@@ -52,15 +55,6 @@ func runFactoryTest(tb testing.TB, test func(tb testing.TB, sw spanstore.Writer,
 	dr, err := f.CreateDependencyReader()
 	assert.NoError(tb, err)
 
-	defer func() {
-		if closer, ok := sw.(io.Closer); ok {
-			err := closer.Close()
-			assert.NoError(tb, err)
-		} else {
-			tb.FailNow()
-		}
-		assert.NoError(tb, f.Close())
-	}()
 	test(tb, sw, dr)
 }
 

--- a/plugin/storage/badger/dependencystore/storage_test.go
+++ b/plugin/storage/badger/dependencystore/storage_test.go
@@ -21,6 +21,7 @@ import (
 	"time"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 	"github.com/uber/jaeger-lib/metrics"
 	"go.uber.org/zap"
 
@@ -35,7 +36,7 @@ import (
 func runFactoryTest(tb testing.TB, test func(tb testing.TB, sw spanstore.Writer, dr dependencystore.Reader)) {
 	f := badger.NewFactory()
 	defer func() {
-		assert.NoError(tb, f.Close())
+		require.NoError(tb, f.Close())
 	}()
 
 	opts := badger.NewOptions("badger")

--- a/plugin/storage/badger/factory.go
+++ b/plugin/storage/badger/factory.go
@@ -158,7 +158,7 @@ func (f *Factory) CreateSpanReader() (spanstore.Reader, error) {
 
 // CreateSpanWriter implements storage.Factory
 func (f *Factory) CreateSpanWriter() (spanstore.Writer, error) {
-	return badgerStore.NewSpanWriter(f.store, f.cache, f.Options.Primary.SpanStoreTTL, f), nil
+	return badgerStore.NewSpanWriter(f.store, f.cache, f.Options.Primary.SpanStoreTTL), nil
 }
 
 // CreateDependencyReader implements storage.Factory

--- a/plugin/storage/badger/spanstore/read_write_test.go
+++ b/plugin/storage/badger/spanstore/read_write_test.go
@@ -415,7 +415,7 @@ func TestPersist(t *testing.T) {
 			} else {
 				t.FailNow()
 			}
-
+			assert.NoError(t, f.Close())
 		}()
 
 		test(t, sw, sr)

--- a/plugin/storage/badger/spanstore/read_write_test.go
+++ b/plugin/storage/badger/spanstore/read_write_test.go
@@ -17,7 +17,6 @@ package spanstore_test
 import (
 	"context"
 	"fmt"
-	"io"
 	"io/ioutil"
 	"log"
 	"math/rand"
@@ -27,17 +26,15 @@ import (
 	"time"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 	"github.com/uber/jaeger-lib/metrics"
 	"go.uber.org/zap"
 
 	"github.com/jaegertracing/jaeger/model"
 	"github.com/jaegertracing/jaeger/pkg/config"
 	"github.com/jaegertracing/jaeger/plugin/storage/badger"
-	bss "github.com/jaegertracing/jaeger/plugin/storage/badger/spanstore"
 	"github.com/jaegertracing/jaeger/storage/spanstore"
 )
-
-var _ io.Closer = new(bss.SpanWriter)
 
 func TestWriteReadBack(t *testing.T) {
 	runFactoryTest(t, func(tb testing.TB, sw spanstore.Writer, sr spanstore.Reader) {
@@ -385,6 +382,10 @@ func TestPersist(t *testing.T) {
 
 	p := func(t *testing.T, dir string, test func(t *testing.T, sw spanstore.Writer, sr spanstore.Reader)) {
 		f := badger.NewFactory()
+		defer func() {
+			require.NoError(t, f.Close())
+		}()
+
 		opts := badger.NewOptions("badger")
 		v, command := config.Viperize(opts.AddFlags)
 
@@ -407,16 +408,6 @@ func TestPersist(t *testing.T) {
 
 		sr, err := f.CreateSpanReader()
 		assert.NoError(t, err)
-
-		defer func() {
-			if closer, ok := sw.(io.Closer); ok {
-				err := closer.Close()
-				assert.NoError(t, err)
-			} else {
-				t.FailNow()
-			}
-			assert.NoError(t, f.Close())
-		}()
 
 		test(t, sw, sr)
 	}
@@ -456,6 +447,10 @@ func TestPersist(t *testing.T) {
 // Opens a badger db and runs a test on it.
 func runFactoryTest(tb testing.TB, test func(tb testing.TB, sw spanstore.Writer, sr spanstore.Reader)) {
 	f := badger.NewFactory()
+	defer func() {
+		require.NoError(tb, f.Close())
+	}()
+
 	opts := badger.NewOptions("badger")
 	v, command := config.Viperize(opts.AddFlags)
 	command.ParseFlags([]string{
@@ -473,15 +468,6 @@ func runFactoryTest(tb testing.TB, test func(tb testing.TB, sw spanstore.Writer,
 	sr, err := f.CreateSpanReader()
 	assert.NoError(tb, err)
 
-	defer func() {
-		if closer, ok := sw.(io.Closer); ok {
-			err := closer.Close()
-			assert.NoError(tb, err)
-		} else {
-			tb.FailNow()
-		}
-
-	}()
 	test(tb, sw, sr)
 }
 
@@ -654,13 +640,9 @@ func runLargeFactoryTest(tb testing.TB, test func(tb testing.TB, sw spanstore.Wr
 	assert.NoError(err)
 
 	defer func() {
-		if closer, ok := sw.(io.Closer); ok {
-			err := closer.Close()
-			os.RemoveAll(dir)
-			assert.NoError(err)
-		} else {
-			assert.FailNow("io.Closer not implemented by SpanWriter")
-		}
+		err := f.Close()
+		os.RemoveAll(dir)
+		require.NoError(tb, err)
 	}()
 	test(tb, sw, sr)
 }

--- a/plugin/storage/badger/spanstore/rw_internal_test.go
+++ b/plugin/storage/badger/spanstore/rw_internal_test.go
@@ -33,7 +33,7 @@ func TestEncodingTypes(t *testing.T) {
 		testSpan := createDummySpan()
 
 		cache := NewCacheStore(store, time.Duration(1*time.Hour), true)
-		sw := NewSpanWriter(store, cache, time.Duration(1*time.Hour), nil)
+		sw := NewSpanWriter(store, cache, time.Duration(1*time.Hour))
 		rw := NewTraceReader(store, cache)
 
 		sw.encodingType = jsonEncoding
@@ -50,7 +50,7 @@ func TestEncodingTypes(t *testing.T) {
 		testSpan := createDummySpan()
 
 		cache := NewCacheStore(store, time.Duration(1*time.Hour), true)
-		sw := NewSpanWriter(store, cache, time.Duration(1*time.Hour), nil)
+		sw := NewSpanWriter(store, cache, time.Duration(1*time.Hour))
 		// rw := NewTraceReader(store, cache)
 
 		sw.encodingType = 0x04
@@ -63,7 +63,7 @@ func TestEncodingTypes(t *testing.T) {
 		testSpan := createDummySpan()
 
 		cache := NewCacheStore(store, time.Duration(1*time.Hour), true)
-		sw := NewSpanWriter(store, cache, time.Duration(1*time.Hour), nil)
+		sw := NewSpanWriter(store, cache, time.Duration(1*time.Hour))
 		rw := NewTraceReader(store, cache)
 
 		err := sw.WriteSpan(context.Background(), &testSpan)
@@ -102,7 +102,7 @@ func TestDuplicateTraceIDDetection(t *testing.T) {
 	runWithBadger(t, func(store *badger.DB, t *testing.T) {
 		testSpan := createDummySpan()
 		cache := NewCacheStore(store, time.Duration(1*time.Hour), true)
-		sw := NewSpanWriter(store, cache, time.Duration(1*time.Hour), nil)
+		sw := NewSpanWriter(store, cache, time.Duration(1*time.Hour))
 		rw := NewTraceReader(store, cache)
 		origStartTime := testSpan.StartTime
 

--- a/plugin/storage/badger/spanstore/writer.go
+++ b/plugin/storage/badger/spanstore/writer.go
@@ -189,8 +189,3 @@ func createTraceKV(span *model.Span, encodingType byte, startTime uint64) ([]byt
 
 	return key, bb, err
 }
-
-// Close Implements io.Closer but does nothing.
-func (w *SpanWriter) Close() error {
-	return nil
-}

--- a/plugin/storage/badger/spanstore/writer.go
+++ b/plugin/storage/badger/spanstore/writer.go
@@ -19,7 +19,6 @@ import (
 	"encoding/binary"
 	"encoding/json"
 	"fmt"
-	"io"
 	"time"
 
 	"github.com/dgraph-io/badger"
@@ -52,17 +51,15 @@ type SpanWriter struct {
 	store        *badger.DB
 	ttl          time.Duration
 	cache        *CacheStore
-	closer       io.Closer
 	encodingType byte
 }
 
 // NewSpanWriter returns a SpawnWriter with cache
-func NewSpanWriter(db *badger.DB, c *CacheStore, ttl time.Duration, storageCloser io.Closer) *SpanWriter {
+func NewSpanWriter(db *badger.DB, c *CacheStore, ttl time.Duration) *SpanWriter {
 	return &SpanWriter{
 		store:        db,
 		ttl:          ttl,
 		cache:        c,
-		closer:       storageCloser,
 		encodingType: defaultEncoding, // TODO Make configurable
 	}
 }
@@ -193,7 +190,7 @@ func createTraceKV(span *model.Span, encodingType byte, startTime uint64) ([]byt
 	return key, bb, err
 }
 
-// Close Implements io.Closer and closes the underlying storage
+// Close Implements io.Closer but does nothing.
 func (w *SpanWriter) Close() error {
-	return w.closer.Close()
+	return nil
 }

